### PR TITLE
Add improvements to getblock RPC output

### PR DESCRIPTION
**Pull request by bitcartel**
Includes and supercedes #3095. Includes code cherry-picked from bitcoin/bitcoin#8704.

### CODE REVIEW

**Comment by bitcartel** [comment]
This PR exists because I could not push to Peter Pratscher's PR.

File: src/rpcblockchain.cpp [hunk 1]
```diff
@@ -395,12 +395,15 @@ UniValue getblock(const UniValue& params, bool fHelp)
     if (fHelp || params.size() < 1 || params.size() > 2)
         throw runtime_error(
             "getblock \"hash|height\" ( verbose )\n"
```
Comment (str4d) [inline src/rpcblockchain.cpp]: `s/verbose/verbosity/` (lost during the backport). Please rebase this change into the first commit, or into an intermediate commit (so the two together correspond to upstream's commit, and the final commit is our further changes).

File: src/rpcblockchain.cpp [hunk 2]
```diff
@@ -399,7 +399,7 @@ UniValue getblock(const UniValue& params, bool fHelp)
             "If verbosity is 1, returns an Object with information about block <hash>.\n"
             "If verbosity is 2, returns an Object with information about block <hash> and information about each transaction. \n"
             "\nArguments:\n"
-            "1. \"blockhash\"          (string, required) The block hash\n"
+            "1. \"hash|height\"          (string, required) The block hash or height\n"
```
Comment (str4d) [inline src/rpcblockchain.cpp]: Move this into the first commit, or into an intermediate commit (so the two together correspond to upstream's commit, and the final commit is our further changes).

File: src/rpcserver.cpp [hunk 3]
```diff
@@ -27,8 +27,6 @@
 #include <boost/signals2/signal.hpp>
 #include <boost/thread.hpp>
 #include <boost/algorithm/string/case_conv.hpp> // for to_upper()
-#include <boost/algorithm/string/classification.hpp>
-#include <boost/algorithm/string/split.hpp>
```
Comment (str4d) [inline src/rpcserver.cpp]: Move this into the first commit, or into an intermediate commit (so the two together correspond to upstream's commit, and the final commit is our further changes).

File: src/rpcblockchain.cpp [hunk 4]
```diff
@@ -472,15 +490,15 @@ UniValue getblock(const UniValue& params, bool fHelp)
...
+    if (verbosity == 0)
     {
         CDataStream ssBlock(SER_NETWORK, PROTOCOL_VERSION);
         ssBlock << block;
         std::string strHex = HexStr(ssBlock.begin(), ssBlock.end());
         return strHex;
     }
 
-    return blockToJSON(block, pblockindex);
+    return blockToJSON(block, pblockindex, verbosity >= 2);
```
Comment (str4d) [inline src/rpcblockchain.cpp]: Change to `verbosity == 2` to match the earlier change.

Comment (str4d) [reply src/rpcblockchain.cpp]: I see @daira [reasoned it was fine](https://github.com/zcash/zcash/pull/3095#discussion_r176944896); fair enough.

File: src/rpcblockchain.cpp [hunk 5]
```diff
@@ -395,12 +395,15 @@ UniValue getblock(const UniValue& params, bool fHelp)
     if (fHelp || params.size() < 1 || params.size() > 2)
         throw runtime_error(
             "getblock \"hash|height\" ( verbose )\n"
-            "\nIf verbose is false, returns a string that is serialized, hex-encoded data for block 'hash|height'.\n"
-            "If verbose is true, returns an Object with information about block <hash|height>.\n"
+            "\nIf verbosity is 0, returns a string that is serialized, hex-encoded data for block 'hash'.\n"
```
Comment (Eirik0) [inline src/rpcblockchain.cpp]: In several places the documentation says "block 'hash'" or something similar, but this is actually hash or height. Is this something we care to update?

Comment (bitcartel) [reply src/rpcblockchain.cpp]: I tidied up the help which we inherited from upstream.

File: src/rpcserver.cpp (hunk 3 above)
Comment (bitcartel) [reply src/rpcserver.cpp]: Done

File: src/rpcblockchain.cpp (hunk 2 above)
Comment (bitcartel) [reply src/rpcblockchain.cpp]: Done

File: src/rpcblockchain.cpp (hunk 1 above)
Comment (bitcartel) [reply src/rpcblockchain.cpp]: Done

**Comment by bitcartel** [comment]
@str4d @Eirik0 Force pushed changes to address review comments.

**Comment by Eirik0** [comment]
utACK

Looks good

**Comment by str4d** [comment]
@zkbot r+

**Comment by str4d** [comment]
Unrelated transient RPC test failure.

@zkbot retry

**Pull request merged**